### PR TITLE
[target-allocator] marshal with jsoniter

### PR DESCRIPTION
**Pull request by kristinapathak**
https://github.com/open-telemetry/opentelemetry-operator/issues/1336

Using jsoniter to marshal for server responses results in less memory and CPU used.  This can be seen in the benchmarks below, where `std` marks the previous code and `jsoniter` shows the performance with this PR.

### CODE REVIEW

**Comment by kristinapathak** [comment]
benchmark results: 
[benchmarks.txt](https://github.com/open-telemetry/opentelemetry-operator/files/10273599/benchmarks.txt)

File: cmd/otel-allocator/server/server.go [hunk 1]
```diff
@@ -42,27 +41,49 @@ var (
...
+}
+
 type Server struct {
 	logger           logr.Logger
 	allocator        allocation.Allocator
-	discoveryManager *target.Discoverer
+	discoveryManager DiscoveryManager
 	server           *http.Server
+	yamlMarshaller   jsoniter.API // TODO: for testing, it would make a lot of sense to have a simpler interface
+	jsonMarshaller   jsoniter.API // TODO: for testing, it would make a lot of sense to have a simpler interface
```
Comment (kristinapathak) [inline cmd/otel-allocator/server/server.go:73]: @jaronoff97, do you think I should define my own interfaces here or keep using the `jsoniter.API` interface (which is pretty big)?

Comment (jaronoff97) [reply cmd/otel-allocator/server/server.go:73]: Is there a more generic interface these adhere to?

File: cmd/otel-allocator/server/server.go [hunk 2]
```diff
@@ -42,27 +41,49 @@ var (
...
+		ObjectFieldMustBeSimpleString: true,
+	}.Froze()
+)
+
 type collectorJSON struct {
 	Link string         `json:"_link"`
 	Jobs []*target.Item `json:"targets"`
 }
 
+type DiscoveryManager interface {
```
Comment (jaronoff97) [inline cmd/otel-allocator/server/server.go:63]: nice add here

File: cmd/otel-allocator/server/server_test.go [hunk 3]
```diff
@@ -15,15 +15,16 @@
 package server
 
 import (
-	"crypto/rand"
 	"encoding/json"
 	"fmt"
 	"io"
-	"math/big"
+	"math/rand"
```
Comment (jaronoff97) [inline cmd/otel-allocator/server/server_test.go:21]: I thought we had a linter that discourages the use of `math/rand` in favor of `crypto/rand`, i guess it's fine given this is just a test.

Comment (jaronoff97) [reply cmd/otel-allocator/server/server_test.go:21]: oh now i see, you're disabling the linter down there, that's fine

**Review by jaronoff97 — APPROVED** [review]

File: cmd/otel-allocator/server/server.go (hunk 1 above)
Comment (kristinapathak) [reply cmd/otel-allocator/server/server.go:73]: Not really 😕

**Pull request merged**